src/libproc_macro/bridge.rs

```diff
@@ -0,0 +1,856 @@
+                    &storage.$oty[self.0]
+                }
+            }
+
+            impl<S: server::Types> HandleToData<'s, S> for &'s mut $oty {
+                type Data = &'s mut S::$oty;
+                fn to_data(self, storage: &'s mut Storage<S>) -> Self::Data {
+                    &mut storage.$oty[self.0]
+                }
+            }
+        )*
+
+        $(
+            #[repr(C)]
+            #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+            pub(crate) struct $ity(storage::Handle);
+            impl !Send for $ity {}
+            impl !Sync for $ity {}
+
+            impl<S: server::Types> HandleFromData<S> for $ity {
+                type Data = S::$ity;
+                fn from_data(x: Self::Data, storage: &mut Storage<S>) -> Self {
+                    $ity(storage.$ity.alloc(x))
+                }
+            }
+
+            impl<S: server::Types> HandleToData<'_, S> for $ity {
+                type Data = S::$ity;
+                fn to_data(self, storage: &mut Storage<S>) -> Self::Data {
+                    storage.$ity.take(self.0)
+                }
+            }
+        )*
+    }
+}
+define_handles! {
+    'owned:
+    TokenStream,
+    TokenStreamBuilder,
+    TokenStreamIter,
+    Group,
+    Literal,
+    SourceFile,
+    Diagnostic,
+
+    'interned:
+    Op,
+    Term,
+    Span,
+}
+
+// FIXME(eddyb) generate these impls by pattern-matching on the
+// names of methods - also could use the presence of `fn drop`
+// to distinguish between 'owned and 'interned, above.
+// Alternatively, special 'modes" could be listed of types in with_api
+// instead of pattern matching on methods, here and in server decl.
+
+impl Clone for TokenStream {
+    fn clone(&self) -> Self {
+        self.clone()
+    }
+}
+
+impl fmt::Display for TokenStream {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(&self.display())
+    }
+}
+
+impl Clone for TokenStreamIter {
+    fn clone(&self) -> Self {
+        self.clone()
+    }
+}
+
+impl Clone for Group {
+    fn clone(&self) -> Self {
+        self.clone()
+    }
+}
+
+impl Clone for Literal {
+    fn clone(&self) -> Self {
+        self.clone()
+    }
+}
+
+// FIXME(eddyb) `Literal` should not expose internal `Debug` impls.
+impl fmt::Debug for Literal {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(&self.debug())
+    }
+}
+
+impl Clone for SourceFile {
+    fn clone(&self) -> Self {
+        self.clone()
+    }
+}
+
+impl fmt::Debug for Span {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(&self.debug())
+    }
+}
+
+macro_rules! define_client_side {
+    ($($name:ident {
+        $(fn $method:ident($($arg:ident: $arg_ty:ty),* $(,)?) $(-> $ret_ty:ty)?;)*
+    }),* $(,)?) => {
+        // FIXME(eddyb) generate a manual C ABI vtable instead.
+        type Server<'a> = server::Server<$($name = $name),*> + 'a;
```

Comment:
This is an example of one I'm a little worried about. If we change the compiler to tweak vtable layout, we *could* switch this to a C ABI instead but that'd be a huge pain and would probably prevent such a change from landing in the compiler, right?

Reply:
I think this is one big one I'd prefer to explicitly make ABI compatible before we land this PR, this definitely seems to be the lynchpin!

Reply:
Yeah, I definitely want to try and switch this to C ABI - the main annoyance is having to use generative typing to enforce that the data pointer is not misused.

Reply:
Note that there's the same problem with `Registry` too, and hopefully I can reuse the same solution for both.